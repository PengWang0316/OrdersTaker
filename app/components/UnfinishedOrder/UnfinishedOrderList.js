import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  List, ListItem, ListItemText, ListItemSecondaryAction, Avatar, Typography
} from '@material-ui/core';
import { CheckCircle as CheckCircleIcon } from '@material-ui/icons';
import { withStyles } from '@material-ui/core/styles';
import { lightGreen } from '@material-ui/core/colors';

import { ITEM_ONE_PRICE_KEY } from '../../config';

const styles = {
  root: {
    width: '100%'
  },
  avatar: {
    marginRight: 10
  },
  checkIcon: {
    color: lightGreen[500]
  }
};

const getQtyString = qtyObject => {
  let returnStr = '';
  Object.keys(qtyObject).forEach(key => {
    if (key !== ITEM_ONE_PRICE_KEY) returnStr += `${key} × ${qtyObject[key]} `;
    else returnStr += `× ${qtyObject[key]}`;
  });
  return returnStr;
};

export const UnfinishedOrderList = ({
  order, menuItems, classes, onClick
}) => (
  <List className={classes.root}>
    {menuItems && Object.keys(order.items).map(key => (
      <ListItem key={key} button name={key} onClick={() => onClick(order._id, key)}>
        <Avatar src={menuItems[key].photo} className={classes.avatar} />
        <Typography color="primary">
          {menuItems[key].name}
        </Typography>
        <ListItemText primary={getQtyString(order.items[key].qty)} />
        {order.finishedItems && order.finishedItems[key] && (
          <ListItemSecondaryAction>
            <CheckCircleIcon className={classes.checkIcon} />
          </ListItemSecondaryAction>
        )}
      </ListItem>
    ))}
  </List>
);
UnfinishedOrderList.propTypes = {
  classes: PropTypes.object.isRequired,
  order: PropTypes.object.isRequired,
  menuItems: PropTypes.object,
  onClick: PropTypes.func.isRequired
};
UnfinishedOrderList.defaultProps = {
  menuItems: null
};
/* istanbul ignore next */
const mapStateToProps = state => ({
  menuItems: state.menuItems
});
export default connect(mapStateToProps, null)(withStyles(styles)(UnfinishedOrderList));
