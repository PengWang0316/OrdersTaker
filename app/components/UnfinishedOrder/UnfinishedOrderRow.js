import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Typography, CircularProgress } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { DoneAll as DoneAllIcon } from '@material-ui/icons';
import { lightGreen } from '@material-ui/core/colors';

const styles = {
  root: {
    width: '100%',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  flexDiv: {
    display: 'flex',
    alignItems: 'center'
  },
  finishPercentage: {
    fontWeight: 'bold',
    size: 14
  },
  progress: {
    position: 'relative',
    left: -33
  },
  circleIcon: {
    width: 30,
    height: 30,
    color: lightGreen[700]
  }
};

const getPercentageOfFinishing = (finishedItemsLength, itemsLength) => Math.floor((finishedItemsLength / itemsLength) * 100);

export const UnfinishedOrderRow = ({ order, classes }) => {
  const orderDate = new Date(order.dateStamp);
  let finishingPercentage = 0;
  if (order.finishedItems) { // If the order has some finished items, calculate the percentage for the finishing.
    const finishedItemsLength = Object.keys(order.finishedItems).length;
    const itemsLength = Object.keys(order.items).length;
    finishingPercentage = finishedItemsLength === itemsLength
      ? 100 : getPercentageOfFinishing(finishedItemsLength, itemsLength);
  }
  return (
    <div className={classes.root}>
      <Typography color="primary">Table {order.tableNumber}</Typography>
      <Typography color="textPrimary">
        {`${orderDate.getHours()}:${orderDate.getMinutes()} ${orderDate.toDateString()}`}
      </Typography>
      <div className={classes.flexDiv}>
        {finishingPercentage === 100 && <DoneAllIcon className={`${classes.circleIcon} ${classes.progress}`} />}
        {finishingPercentage !== 100 && (
          <Fragment>
            <Typography color="primary" className={classes.finishPercentage}>{finishingPercentage}%</Typography>
            <CircularProgress variant="static" value={finishingPercentage} size={36} className={classes.progress} />
          </Fragment>
        )}
      </div>
    </div>
  );
};
UnfinishedOrderRow.propTypes = {
  classes: PropTypes.object.isRequired,
  order: PropTypes.object.isRequired
};
export default withStyles(styles)(UnfinishedOrderRow);
