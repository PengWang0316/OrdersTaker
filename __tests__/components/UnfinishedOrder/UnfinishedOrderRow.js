import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { UnfinishedOrderRow } from '../../../app/components/UnfinishedOrder/UnfinishedOrderRow';

jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/core/CircularProgress', () => 'CircularProgress');
jest.mock('@material-ui/icons/DoneAll', () => 'DoneAll');
jest.mock('@material-ui/icons/Delete', () => 'Delete');

describe('UnfinishedOrderRow', () => {
  const defaultProps = {
    classes: {
      root: 'root',
      flexDiv: 'flexDiv',
      finishPercentage: 'finishPercentage',
      progress: 'progress',
      circleIcon: 'circleIcon'
    },
    order: {
      dateStamp: '2018-07-27T22:22:53.647Z'
    },
    deleteOrderCallback: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<UnfinishedOrderRow {...props} />);

  test('Snapshot without finishedItems', () => expect(renderer.create(<UnfinishedOrderRow {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot with finishedItems and different length', () => expect(renderer.create(<UnfinishedOrderRow {...{ ...defaultProps, order: { items: { a: true, b: true }, finishedItems: { a: true }, dateStamp: '2018-07-27T22:22:53.647Z' } }} />).toJSON()).toMatchSnapshot());
  test('Snapshot with finishedItems and same length', () => expect(renderer.create(<UnfinishedOrderRow {...{ ...defaultProps, order: { items: { a: true, b: true }, finishedItems: { a: true, b: true }, dateStamp: '2018-07-27T22:22:53.647Z' } }} />).toJSON()).toMatchSnapshot());

  test('deleteOrderCallback', () => {
    const component = getShallowComponent({ ...defaultProps, order: { _id: 'orderId', items: { a: true, b: true }, finishedItems: { a: true, b: true }, dateStamp: '2018-07-27T22:22:53.647Z' } });
    component.find('IconButton').simulate('click');
    expect(defaultProps.deleteOrderCallback).toHaveBeenCalledTimes(1);
    expect(defaultProps.deleteOrderCallback).toHaveBeenLastCalledWith('orderId');
  });
});
