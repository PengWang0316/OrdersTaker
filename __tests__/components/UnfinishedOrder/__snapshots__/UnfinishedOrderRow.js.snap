// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnfinishedOrderRow Snapshot with finishedItems and different length 1`] = `
<div
  className="root"
>
  <Typography
    color="primary"
  >
    Table 
  </Typography>
  <Typography
    color="textPrimary"
  >
    15:22 Fri Jul 27 2018
  </Typography>
  <div
    className="flexDiv"
  >
    <Typography
      className="finishPercentage"
      color="primary"
    >
      50
      %
    </Typography>
    <CircularProgress
      className="progress"
      size={36}
      value={50}
      variant="static"
    />
  </div>
</div>
`;

exports[`UnfinishedOrderRow Snapshot with finishedItems and same length 1`] = `
<div
  className="root"
>
  <Typography
    color="primary"
  >
    Table 
  </Typography>
  <Typography
    color="textPrimary"
  >
    15:22 Fri Jul 27 2018
  </Typography>
  <div
    className="flexDiv"
  >
    <DoneAll
      className="circleIcon progress"
    />
  </div>
</div>
`;

exports[`UnfinishedOrderRow Snapshot without finishedItems 1`] = `
<div
  className="root"
>
  <Typography
    color="primary"
  >
    Table 
  </Typography>
  <Typography
    color="textPrimary"
  >
    15:22 Fri Jul 27 2018
  </Typography>
  <div
    className="flexDiv"
  >
    <Typography
      className="finishPercentage"
      color="primary"
    >
      0
      %
    </Typography>
    <CircularProgress
      className="progress"
      size={36}
      value={0}
      variant="static"
    />
  </div>
</div>
`;
